### DISCUSSION TITLE
render PlantUML / Mermaid / ... in Freeplane nodes

### DISCUSSION THREAD

**Discussion by nnako** [freeplane/freeplane#2082] [Ideas]
Hi folks,

there seems to be growing interest in adding further features to Freeplane which enhance it's ability to render content in a more presentable way than "just" showing (even styled) textual nodes and scrolling / unfolding through the mindmap.

As mentioned in a [question](https://github.com/freeplane/freeplane/discussions/1570) by @biswajit-goswami, there might even be an option to integrate these further visualization features into existing markdown renderers. And @Coffee-Search developed [very impressive representations](https://github.com/freeplane/freeplane/discussions/254#discussioncomment-10970606) using Freeplane's current styling features combined with the "free floating node" concept:

**e.g. weekly planner** (yes, this is displayed and can be modified directly within the Freeplane editor)
![image](https://github.com/user-attachments/assets/94cf4db5-a363-4aed-955e-dfd2bd5f3030)

**e.g. sequence diagram** (yes, this is displayed and can be modified directly within the Freeplane editor)
![image](https://github.com/user-attachments/assets/4b1f1f3a-0a05-47c3-b1e5-f1d6fde65862)

The common problem with using these means on board seems to be:

- the user interface within Freeplane editor is not yet suited to comfortably edit these diagrams manually, as even selecting the desired element for modification might take its time, let alone changing sizes and style.
- semi-automatic (or at least simplified) creation of these diagrams from external task management systems or even code bases does not yet seem to be possible.

I'd like to start the development of a renderer (like already done by some fine devs with the ability to direktly input `markdown` syntax into nodes and let Freeplane render their output) for [`PlantUML`](https://plantuml.com/) or [`Mermaid`](https://mermaid.js.org/). Those are wonderful resources when it comes to displaying technical information in an appealing way while providing the ability to change the content on the fly in a textual manner and even use text-based version management for the contained content and its presentation.

Are there any devs like @dpolivaev, @fnatter, @vboerchers, @ldmpub, @macmarrum, ... who would would be interested in such a feature and like to join and support finding an adequate spot within Freeplane's code base (maybe as a plugin) and the right concept for integration of such a renderer accessing an external resource?

Looking forward to seeing these things happen... ;-)

**Comment by macmarrum** [comment]
I've had experience with PlantUML in GitLab, where the diagram definition is part of Markdown. This  makes it possible to add content before or after the diagram or include many diagrams, one after another – https://docs.gitlab.com/ee/administration/integration/plantuml.html

I know that Freeplane uses Markedj for Markdown. In Markedj's GitHub page I can see a section about extensions – https://github.com/gitbucket/markedj?tab=readme-ov-file#extensions

I propose to go this way, i.e. to create a PlantUML extension for Markedj and use it in Freeplane.

The extension would find the **plantuml** section(s)
~~~
```plantuml
Bob -> Alice : hello
Alice -> Bob : hi
```
~~~
and convert it to PNG, encode it as base64 and place it in `<img src="data:image/png;base64,...">`. This would display the diagram as an image, inside Markdown content (in node core or details or note).

As to an actual implementation, I'm not sure how much time I'll be able to devote to it. So if you think this is a good idea and have time, please feel free to start coding.

**Comment by macmarrum** [comment]
I've added PlantUML (https://plantuml.com) support to Markdown in Freeplane -> https://github.com/freeplane/freeplane/commit/e602267e07b2287b7a9e84c5503887faf732c6bd

If no objections from @dpolivaev, it will get included in the next preview.

The initial idea to use the GitLab way of indicating PlantUML code is imcompatible with the current implementation of Markedj (used by Freeplane for Markdown).

I've opted for using the native PlantUML tags, i.e. 
- `@startuml` /  `@enduml`
- `@startgantt` /  `@endgantt`
- `@startchronology` /  `@endchronology`
- `@startsalt` /  `@endsalt`
- `@startjson` /  `@endjson`
- `@startyaml` /  `@endyaml`
- `@startebnf` /  `@endebnf`
- `@startregex` /  `@endregex`
- `@startwbs` /  `@endwbs`
- `@startmindmap` /  `@endmindmap`

### Examples

_Can be copied and pasted to Markdown nodes in Freeplane_

```
@startuml
title This is a Use Case diagram

:Actor: --> (Use Case)
@enduml
```

```
@startgantt
title This is a Gantt Chart

[T1] requires 4 days
[T2] requires 4 days and starts 3 days after [T1]'s end with blue dotted link
[T3] requires 4 days and starts 3 days after [T2]'s end with green bold link
[T4] requires 4 days and starts 3 days after [T3]'s end with green dashed link
@endgantt
```

**Reply by dpolivaev** [reply]
It adds 12 MB to every distribution.

**Reply by macmarrum** [reply]
PlantUML is powerful, but apparently a bit heavy.

One idea is to have a user download and put plantuml.jar in the classpath of Freeplane, if they want to use PlantUML.

**Reply by dpolivaev** [reply]
If it works for you I think it would be nicer. How do you see it?

**Reply by macmarrum** [reply]
Sure. I don't know yet how to do it, but I will try.

**Reply by dpolivaev** [reply]
If you know the jar file name and path, you can check if it is available and create your own URLClassLoader that knows this URL and MarkdownRenderer.class.getClassLoader() as a parent. You  can use it to load class  PlantUMLExtension and call markdownRenderer.addExtension(Extention extension) to register it.

**Reply by dpolivaev** [reply]
Then users do not need to add it to Freeplane classpath and can download it and put it into their user configuration directory.

**Reply by dpolivaev** [reply]
You can use ResourceController.getResource(String) to look for the resource in both user directory and installation directory

**Reply by macmarrum** [reply]
Thanks for the hints :+1:

It will take me some time (might be days or even weeks) to learn more about this topic.

If you have time and are willing, please feel free to introduce the necessary changes. 
And if you want to create a release without PlantUML before the externalization of plantuml.jar is implemented, you can revert the commit.

**Reply by dpolivaev** [reply]
I can implement it, it seems easy enough.

**Reply by dpolivaev** [reply]
About 10 lines of code

**Reply by dpolivaev** [reply]
I implemented it, but it was more difficult than I originally thought.

**Reply by macmarrum** [reply]
Fantastic! :rocket: Thank you so much!

**Comment by macmarrum** [comment]
I've noticed that Freeplane renders the Markdown of a node when the node's selection changes. 
In case of PlantUML content, PNG is generated each time a node is selected or unselected.

Here's a diagram that shows the current time (as of rendering). It can be used to observe this behavior.

```
@startuml
!$dt = %date("HH:mm:ss", %now())
($dt)
@enduml
```

**Reply by dpolivaev** [reply]
Improved

**Reply by dpolivaev** [reply]
https://github.com/freeplane/freeplane/issues/2201

**Reply by dpolivaev** [reply]
https://github.com/freeplane/freeplane/commit/0867c38c68524291971e9fe9992973b5f0fc9a89

**Reply by macmarrum** [reply]
This is some good magic :magic_wand:
Very cool :sunglasses:
Thank you

**Reply by dpolivaev** [reply]
I replaced it by another fix because displayed text can depend on Freeplane properties and node formatting including styles. At the moment I specifically avoid recalculation on selection change, but not on some other occasions.

**Reply by macmarrum** [reply]
:+1:

The fact that Markdown rendering isn't triggered on node (un)selection is a huge improvement :rocket: 

I'm going to have a think about other situations which don't require Markdown to be rendered. Maybe there is some more room for optimization.

**Comment by dpolivaev** [comment]
Preview 1.12.9_04

  - to activate it, download recent plantuml jar
  - and save it as `<user-directory>/resources/plantuml.jar`

**Comment by Illegit** [comment]
Very nice. I have a question: Is it possible to build the command of plantuml like a branch or as a sort of modularity?
Here below the branch example
![Screenshot 2025-01-09 225154](https://github.com/user-attachments/assets/4a853125-1c5e-4d9d-b81a-10e6c1ecc570)
here you have the command and the red arrow is showing the node result, but what about having something as in the blue cloud that is creating the same result as for the red arrow in a node (maybe ht result in a different node). So I'm looking at how to introduce calculation in the gantt value instead of writing just the code... My question because I do not know plantuml and its interaction with FP nodes and branches.

**Reply by macmarrum** [reply]
You can use a formula to build your Markdown code.

Example

Copy and past the following nodes into Freeplane

```
Paste the formula in this node
	Analysis
		requires
			4
				days
	Devel
		requires
			4
				days
		and
			starts
				1
					week
						after
							Analysis
								start
		with
			blue dotted link
	SIT
		requires
			4
				days
		and
			starts
				1
					days
						before
							Devel
								end
		with
			green bold link
	UAT
		requires
			4
				days
		and
			starts
				3
					day
						after
							SIT
								end
		with
			green dashed link
```

and replace the branch root's core with the below formula

```groovy
=
def list = ['@startgantt\n']
node.children.each { child ->
    list << "[${child.text}] "
    def listOfGrandchildrenText = child.findAll().drop(1)*.text
    // replace `???` with `[???]'s` if followed by `end` or `start`
    ['end', 'start'].each { end_or_start ->
        if (end_or_start in listOfGrandchildrenText) {
            def i = listOfGrandchildrenText.indexOf(end_or_start)
            listOfGrandchildrenText[i-1] = "[${listOfGrandchildrenText[i-1]}]'s"
        }
    }
    list << listOfGrandchildrenText.join(' ')
    list << '\n'
}
list << '@endgantt'
return list.join('')
```

Finally, for the node with the formula, set format to Markdown.

**Reply by Illegit** [reply]
Nice! Thank you.

**Reply by Illegit** [reply]
After created nodes hierarchy, I tried to insert a formula linking to an external node for a number (e.g: last number 3 in the above example), but in this case the formula has failed because it is reading just the formula and not the result. Is there an easy trick to solve it?

**Reply by macmarrum** [reply]
Yes. Use `transformedText` in place of `text`, i.e.

```groovy
=
def list = ['@startgantt\n']
node.children.each { child ->
    list << "[${child.transformedText}] "
    def listOfGrandchildrenText = child.findAll().drop(1)*.transformedText
    // replace `???` with `[???]'s` if followed by `end` or `start`
    ['end', 'start'].each { end_or_start ->
        if (end_or_start in listOfGrandchildrenText) {
            def i = listOfGrandchildrenText.indexOf(end_or_start)
            listOfGrandchildrenText[i-1] = "[${listOfGrandchildrenText[i-1]}]'s"
        }
    }
    list << listOfGrandchildrenText.join(' ')
    list << '\n'
}
list << '@endgantt'
return list.join('')
```

**Reply by macmarrum** [reply]
Some more improvements to the script. 
Must use an icon — "black square button" by default — to indicate a task node.
Can use an icon — "cross mark" by default — to ignore a node or branch.

```groovy
=import org.freeplane.plugin.script.proxy.ConversionException

skipIconName = 'emoji-274C' // cross mark
taskIconName = 'emoji-1F532' // black square button

def isSkip(n) {
    n.icons.contains(skipIconName)
}

def isTask(n) {
    n.icons.contains(taskIconName)
}

def txt(n) {
    if (isTask(n)) {
        return "[${n.transformedText}]"
    }
    try {
        return n.to.date.format("yyyy-MM-dd")
    } catch (ConvertionException) {
        return n.transformedText
    }
}

def list = ['@startgantt\n']
node.children.findAll { !isSkip(it) }.each { child ->
    def listOfGrandchildrenText = child.find { !isSkip(it) }.collect { txt(it) }
    // replace `???` with `???'s` if followed by `end` or `start`
    ['end', 'start'].each { end_or_start ->
        def i = listOfGrandchildrenText.indexOf(end_or_start)
        if (i > -1) { // i.e. end_or_start in listOfGrandchildrenText
            listOfGrandchildrenText[i-1] = "${listOfGrandchildrenText[i-1]}'s"
        }
    }
    list << listOfGrandchildrenText.join(' ')
    list << '\n'
}
list << '@endgantt'
return list.join('')
```

And here are some nodes to go with it

```
Paste the formula in this node
	project starts
		2025-01-04
	saturdays are closed
	sundays are closed
	Analysis
		requires
			4
				days
	Devel
		requires
			4
				days
		and
			starts
				1
					working
						week
							after
								Analysis
									start
		with
			blue dotted link
	SIT
		requires
			4
				days
		and
			starts
				1
					days
						before
							Devel
								end
		with
			green bold link
	UAT
		requires
			4
				days
		and
			starts
				3
					working
						day
							after
								SIT
									end
		with
			green dashed link

```

**Comment by Illegit** [comment]
Hi nice, but the last script is failing something:
if I activate just the 1st branch it is ok:
![Screenshot 2025-01-10 174555](https://github.com/user-attachments/assets/52c601d2-5952-4ddf-8c13-6c23a5235352)

but when I activate the 2nd it is failing: 
![Screenshot 2025-01-10 174657](https://github.com/user-attachments/assets/f152c33d-7610-4087-82d4-6b4a0abcdcd3)

As you can see I used icons "zzz" for pausing a branch and "." for defining a task

**Reply by macmarrum** [reply]
To see what code is generated, set node format to Standard, for the formula (diagram) node.

It looks to me that you are missing a "." icon for your other task nodes, in particular "after **Analysis** start"

BTW, the script will also work if you use the native PlantUML syntax for tasks, i.e. "[Analysis]" but without the icon.

**Reply by Illegit** [reply]
Hi macmarrum, please observe the examples
1) just square bracket: fails
![Screenshot 2025-01-10 232444](https://github.com/user-attachments/assets/5bddb8dc-4dc8-4166-9cf1-4eb87871057f)
2) square bracket only for 1st task: OK
![Screenshot 2025-01-10 232700](https://github.com/user-attachments/assets/5c97def2-69e3-4c6c-9d50-a3f4ef2175d0)

There is the same behavior as with the (.) icon...

script (with some added lines, that do not change the problem) and example are the same as above.
Thanks

**Reply by Illegit** [reply]
Macmarrum it seems that the rows that begin with and and with in Devel, SIT and UAT are not read correctly, I eliminated the rows using "with" and simplified the rows using "and" (as in Analysis row) then they are read correctly.

**Reply by macmarrum** [reply]
I can see issue but I am failing in a good description so that you can understand it.

Let me try another approach. In Freeplane, search you mind map for nodes with text "Analysis". Each such node is a task, threfore it should have the `.` icon or be surrounded with `[]`.
The same goes for "Devel", "SIT" and "UAT".

If this is still not helpful, attach your mind map, I'll do the changes there, so that you can see.

**Reply by Illegit** [reply]
That I have already done... no way.
here the map
[myMap-Gantt.zip](https://github.com/user-attachments/files/18383045/myMap-Gantt.zip)
Thanks for your help.

**Reply by macmarrum** [reply]
[myMap-Gantt~macmarrum.mm.zip](https://github.com/user-attachments/files/18383072/myMap-Gantt.macmarrum.mm.zip)

If you need anything else, just shout :smile:

**Reply by Illegit** [reply]
AAAH! Today is friday and the week was eavy, better I go to sleep. Many thanks.

**Reply by macmarrum** [reply]
![](https://media.giphy.com/media/dUmiPZhtiqohi/giphy.gif?cid=ecf05e47sye5bdkyimuveicurxt1procb9pq3vpdfhqfp25i&ep=v1_gifs_search&rid=giphy.gif&ct=g)

**Reply by Illegit** [reply]
👌

**Comment by loic-joly-sonarsource** [comment]
I'd just like to say I would love to see mermaid support!

**Reply by macmarrum** [reply]
Mermaid runs on nodejs. Freeplane runs on Java. I'm not sure how these two can be married.

A quick-and-dirty solution is to call **mermaid-cli** from Freeplane.

1. Install [mermaid-cli](https://github.com/mermaid-js/mermaid-cli) and make sure it works from your command line
2. Save [createImageFromMermaid.groovy](https://github.com/macmarrum/freeplane-scripting/raw/refs/heads/main/src/scripts/createImageFromMermaid.groovy) to `<user-dir>/scripts` and restart Freeplane
3. In the script, set the path to `mmdc` (`mmdc.cmd` on Windows) and change any other `settings`, as needed
4. In Freeplane, create a node, put Mermaid definition in its core (or note, or details) and run the script `Mac1->Create->Create Image From Mermaid`
5. The script will create a temporary `.mermaid` file, pass it to `mmdc` to create `.png`, and then add the PNG image to the node

![2025-04-26_18,22,12](https://github.com/user-attachments/assets/5b5cce5b-2cd1-4515-a73c-e0026595dd84)

**Comment by Illegit** [comment]
Hi macmarrum, I use your script a lot for the gantt, but I'd like to have the possibility to group some of the commands, so I ask you if you can modify your script in order to solve my request: 
![image](https://github.com/user-attachments/assets/679dc0ef-6563-4d10-9fb9-761c8c2f7ab2)


in the example the instructions are hidden, but it would be useful to have all of the software instructions below 1 node, and after to have 1 node that is grouping the commands (yellow bubble), and 1 node for each separator grouping its own nodes (1 for each light blue bubble).... so basically to have 2 nodes (settings, commands) + 1 node for each separtor chapter= total 2+n nodes . Thanks

**Reply by macmarrum** [reply]
Hi @Illegit 

You can try this one.

```groovy
=import org.freeplane.plugin.script.proxy.ConversionException

skipIconName = 'emoji-274C' // cross mark
taskIconName = 'emoji-1F532' // black square button

def isSkip(n) {
    n.icons.contains(skipIconName)
}

// get node's text
// - transformed text
// -- additionally surrounded in [] if it's a task
// - ISO date format if it's a date
def txt(n) {
    if (isTask(n)) {
        return "[${n.transformedText}]"
    }
    try {
        return n.to.date.format("yyyy-MM-dd")
    } catch (ConvertionException) {
        return n.transformedText
    }
}

def isTask(n) {
    n.icons.contains(taskIconName)
}

def appendRowContent(sb, child) {
    def listOfGrandchildrenText = child.find { !isSkip(it) }.collect { txt(it) }
    // replace `???` with `???'s` if followed by `end` or `start`
    ['end', 'start'].each { end_or_start ->
        def i = listOfGrandchildrenText.indexOf(end_or_start)
        if (i > -1) { // i.e. end_or_start in listOfGrandchildrenText
            listOfGrandchildrenText[i-1] = "${listOfGrandchildrenText[i-1]}'s"
        }
    }
    sb << listOfGrandchildrenText.join(' ') << '\n'
}

// allow PlantUML root node to be either the digram itself or the node below the diagram
def plantRoot = node.children ? node : node.parent.children[node.parent.getChildPosition(node) + 1]

def sb = new StringBuilder()
plantRoot.children.findAll { !isSkip(it) }.each { firstLevelChild ->
    def firstLevelChildChildren = firstLevelChild.children
    if (firstLevelChildChildren.size() == 1) {
        // not a group header, threfore treat it as a row
        appendRowContent(sb, firstLevelChild)
    } else {
        // a group header, therefore treat each of its children as separate rows
        sb << txt(firstLevelChild) << '\n'
        firstLevelChildChildren.findAll { !isSkip(it) }.each { child -> appendRowContent(sb, child) }
    }
}
def code = sb.toString()
return (code =~ /^\n*@startgantt\n/ && code =~ /\n@endgantt\n*$/) ? code : "@startgantt\n$code@endgantt" as String
```

**Comment by Illegit** [comment]
Nice macmarrum, it is working. Quite fully usable. 
Just a note: when using a parent separator (eg: -- Macro2 -- in the above draft), the script accepts 0 children, or 2 or more children, but with only 1 child it shows error. To skip it actually I just add and empty child.
In any case it becomes a very useful script to create gantts. Thank you.

**Comment by macmarrum** [comment]
I've converted the [formula](https://github.com/freeplane/freeplane/discussions/2082#discussioncomment-12723118) into a library and extended it so that it's able to make not only `@startgantt` but also `@startuml` code.

![2025-04-12_13,41,42](https://github.com/user-attachments/assets/16924dc2-5ed2-44f5-beb8-2283a2271eab)

### Installation
1. Download [PlantUml.groovy](https://github.com/macmarrum/freeplane-scripting/raw/refs/heads/main/src/lib/io/github/macmarrum/freeplane/PlantUml.groovy) and save it in `<user-dir>/lib/`, then restart Freeplane
2. In your node, put the formula `=io.github.macmarrum.freeplane.PlantUml.makeGantt(node)` or `=io.github.macmarrum.freeplane.PlantUml.makeUml(node)` and change `Core Text > Format` to **Markdown**
3. Single-clone the node (`Edit->Copy->Copy node (single)`) as many times as you need (`Edit->Paste Clone`)
4. If you want to use different icons, specify them in a hashMap, i.e.

```groovy
=
def settings = [
        skip1: true, // whether to consider the first node a "PlantUml-code parent" and skip it
        noEntryIconName: 'emoji-26D4', // no entry - ignore the entire branch
        ignoredIconName: 'emoji-274C', // cross mark - ignore the node
        sinkIconName: 'emoji-1F300', // cyclone - form one line from the node and all its descendants
        taskIconName: 'emoji-1F532', // black square button - make it a Gantt-chart task by enclosing it in square brackets
]
io.github.macmarrum.freeplane.PlantUml.makeGantt(node, settings) // <- add settings as the second argument
```

### Usage Tips for PlantUML.makeGantt and .makeUml
- `makeGantt` and `makeUml` allow PlantUML-code parent to be either the diagram itself or the next sibling, i.e. the node below the diagram
- A node with more than 1 child is considered its own line, e.g. `-- 1 --` and `-- 2 --` in the screenshot
    - A child can be an empty node -> https://github.com/freeplane/freeplane/discussions/2082#discussioncomment-12726334
- Add `sinkIconName` - **cyclone** by default - to collect the entire branch into a single line

**Comment by Illegit** [comment]
Hi macmarrum, interesting I try now but, what is this hashmap? is it a groovy file? And where I have to place it in order to be read by your PlantUMl.groovy? Or I have to place it in a node or in an attribute or? Thanks

**Reply by macmarrum** [reply]
A hashMap is like a list, but has `key: value` pairs. Here, **settings** is a hashMap:

```groovy
def settings = [
        skip1: true, // whether to consider the first node a "PlantUml-code parent" and skip it
        noEntryIconName: 'emoji-26D4', // no entry - ignore the entire branch
        ignoredIconName: 'emoji-274C', // cross mark - ignore the node
        sinkIconName: 'emoji-1F300', // cyclone - form one line from the node and all its descendants
        taskIconName: 'emoji-1F532', // black square button - make it a Gantt-chart task by enclosing it in square brackets
]
```

Given the example mind map [PlantUml.makeGantt.mm.zip](https://github.com/user-attachments/files/19722035/PlantUml.makeGantt.mm.zip), you only need to edit the formula in `ID_1125607354` and change icon names in `settings`.

![2025-04-12_23,36,48](https://github.com/user-attachments/assets/71fc1740-1536-4433-8bc6-84625dac0c69)

**Comment by Illegit** [comment]
Hi macmarrum. Now it is very clear. Thank you. And compliments for this nice tool! I'm very glad to add this way of creating gantt and contemporarily using the full capabilities of FP.
